lib/backend/k8s/resources/resources.go

```diff
@@ -100,3 +103,81 @@ func SetCalicoMetadataFromK8sAnnotations(calicoRes Resource, k8sRes Resource) {
 		}
 	}
 }
+
+// Store Calico Metadata in the in the k8s resource annotations for CRD backed resources.
+// This should store all Metadata except for those stored in Annotations and Labels and
+// store them in annotations.
+func StoreCalicoMetadataInK8sAnnotations(resIn Resource) (Resource, error) {
+	rom := resIn.GetObjectMeta()
+
+	// Make sure to remove data that is passed to Kubernetes so it is not duplicated in
+	// the annotations.
+	romCopy := &metav1.ObjectMeta{}
+	rom.(*metav1.ObjectMeta).DeepCopyInto(romCopy)
+	romCopy.Name = ""
+	romCopy.Namespace = ""
+	romCopy.ResourceVersion = ""
+	romCopy.Labels = nil
+	romCopy.Annotations = nil
+
+	// Marshal the data and store the json representation in the annotations.
+	metadataBytes, err := json.Marshal(romCopy)
+	if err != nil {
+		return nil, err
+	}
+	annotations := rom.GetAnnotations()
+	if len(annotations) == 0 {
+		annotations = make(map[string]string)
+	}
+	annotations[metadataAnnotation] = string(metadataBytes)
+
+	// Make sure to clear out all of the Calico Metadata out of the ObjectMeta except for
+	// Name, Namespace, ResourceVersion, Labels, and Annotations. Annotations is already
+	// copied so it can be set separately.
+	meta := &metav1.ObjectMeta{}
+	meta.Name = rom.GetName()
+	meta.Namespace = rom.GetNamespace()
+	meta.ResourceVersion = rom.GetResourceVersion()
+	meta.Labels = rom.GetLabels()
+
+	resOut := resIn.DeepCopyObject().(Resource)
+	romOut := resOut.GetObjectMeta()
+	meta.DeepCopyInto(romOut.(*metav1.ObjectMeta))
+	romOut.SetAnnotations(annotations)
+
+	return resOut, nil
+}
+
+// Retrieve all of the Calico Metadata from the k8s resource annotations for CRD backed
+// resources. This should remove the relevant Calico Metadata annotation when it has finished.
+func GetCalicoMetadataFromK8sAnnotations(res Resource) error {
+	rom := res.GetObjectMeta()
+	annotations := rom.GetAnnotations()
+	if len(annotations) == 0 {
```

Comment:
I think we need to be more forgiving about the annotation not being present.  For example, our k8s manifest may want to include some of the CRDs for pre-population.  In that case we should return the labels/annotations/uuid/name/namespace from the resource but nothing else.

Reply:
When we write data through libcalico-go we do want to persist the values that are given to us - so as not to potentially confuse the AAPIS -  but reading stuff is less important because that will include data that has already been configured and the AAPIS will learn about that through its watch.